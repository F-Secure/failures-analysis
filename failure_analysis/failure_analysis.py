from lxml import etree
import glob as glob
import os
import sys
import pandas as pd
import numpy as np
import itertools

from difflib import SequenceMatcher
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import CountVectorizer
import Levenshtein
import jellyfish


def parseXMLs(path):
    failure = []
    testname = []
    filename = []
    classname = []

    path = [f for f in glob.glob(path + "*.xml")]

    for xml in path:
        tree = etree.parse(xml)
        root = tree.getroot()
        if int(root.find("testsuite").attrib["failures"]) > 0:
            for node in root.findall(".//failure"):
                parent = node.getparent()
                # failure.append(node.attrib['message'])
                failure.append(node.text)
                testname.append(parent.attrib["name"])
                filename.append(os.path.basename(xml).lower())
                classname.append(parent.attrib["classname"])

    return failure, testname, filename, classname


# JaccardSimiliarity
# takes in 2 lists and returns jaccard score of the two lists
def jaccardSimilarity(list1, list2):
    s1 = set(list1)
    s2 = set(list2)
    return float(len(s1.intersection(s2)) / len(s1.union(s2)))


# cosineSimVectors
# takes in 2 vectors and returns a cosine similiarity between two vectors
def cosineSimVectors(vec1, vec2):
    vec1 = vec1.reshape(1, -1)
    vec2 = vec2.reshape(1, -1)
    return cosine_similarity(vec1, vec2)[0][0]


# scoreFailures
# takes in a list of tuples of failures and returns 5 lists of 5 different string similiarity algorithms
# sm_ratios, list of similiarity scores between two items in a tuple
# coss, a list of cosine similiarity scores between two items in a tuple
# jaccards, a list of jaccard scores between two items in a tuple
# jaros, a list of jaro scores between two items in a tuple
# levens, a list of levenshtein ratio scores between two items in a tuple


def scoreFailures(failures):
    sm_ratios = []
    coss = []
    jaccards = []
    jaros = []
    levens = []

    for failure in failures:
        str1 = failure[0]
        str2 = failure[1]

        sm = SequenceMatcher(a=str1, b=str2)
        sm_ratios.append(sm.ratio())

        vectorizer = CountVectorizer().fit_transform([str1, str2])
        vectors = vectorizer.toarray()
        cos = cosineSimVectors(vectors[0], vectors[1])
        coss.append(cos)

        jaccard = jaccardSimilarity(str1, str2)
        jaccards.append(jaccard)

        jaro = jellyfish.jaro_distance(str1, str2)
        jaros.append(jaro)

        leven = Levenshtein.ratio(str1, str2)
        levens.append(leven)

    return sm_ratios, coss, jaccards, jaros, levens


def main():
    path = sys.argv[1]

    failure, testname, filename, classname = parseXMLs(path)

    testnames = list(itertools.permutations(testname, 2))
    failures = list(itertools.permutations(failure, 2))
    filenames = list(itertools.permutations(filename, 2))
    classnames = list(itertools.permutations(classname, 2))

    sm_ratios, coss, jaccards, jaros, levens = scoreFailures(failures)

    items = [
        filenames,
        testnames,
        classnames,
        sm_ratios,
        coss,
        jaccards,
        jaros,
        levens,
        failures,
    ]

    dfs = []
    for item in items:
        temp_df = pd.DataFrame(item)
        dfs.append(temp_df)
    df = pd.concat(dfs, axis=1)

    df.columns = [
        "filename1",
        "filename2",
        "testname1",
        "testname2",
        "suitename1",
        "suitename2",
        "sm_ratio",
        "cos",
        "jaccard",
        "jaro",
        "leven",
        "failure1",
        "failure2",
    ]
    df = df[["failure1", "suitename2", "testname2", "filename2", "cos", "leven"]]
    # df = df.sort_values('cos',ascending = False)
    # df = pd.pivot_table(df,index=["failure1","testname2","classname2","filename2"])
    for failure in np.unique(df["failure1"].values):
        print("============== FAILURE START =================")
        print(failure)
        print("============== FAILURE END =================")
        temp = df[df["failure1"] == failure]
        temp = temp.sort_values("cos", ascending=True)
        temp = pd.pivot_table(temp, index=["suitename2", "testname2", "filename2"])
        print(temp)
    # print(df)


if __name__ == "__main__":
    main()
