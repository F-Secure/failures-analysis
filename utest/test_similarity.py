import shutil
import tempfile
from pathlib import Path

import numpy as np
import itertools

import pytest

from failure_analysis.failure_analysis import jaccard_similarity, cosine_sim_vectors, score_failures, run

UTEST_ROOT = Path(__file__).resolve().parent
XUNIT_FILES_DIR = UTEST_ROOT / "resources"
PASS_01_FILE_NAME = "pass_01_.xml"
FAIL_01_FILE_NAME = "failing_01_.xml"
FAIL_02_FILE_NAME = "failing_02_.xml"
PASS_01_FILE_PATH = XUNIT_FILES_DIR / PASS_01_FILE_NAME
FAIL_01_FILE_PATH = XUNIT_FILES_DIR / FAIL_01_FILE_NAME
FAIL_02_FILE_PATH = XUNIT_FILES_DIR / FAIL_02_FILE_NAME
EXPECTED_OUTPUT_START = """============== FAILURE START =================
def test_02():
>       assert False
E       assert False


"""

EXPECTED_OUTPUT_END = """============== FAILURE END =================
                                         cos  leven
suitename2    testname2 filename2                  
tests.test_me test_02   failing_01_.xml  1.0    1.0
                        failing_02_.xml  1.0    1.0
                        failing_03_.xml  1.0    1.0
"""  # noqa: W291


def test_jaccard_similarity():
    list1 = ["one", "two"]
    list2 = ["one", "two"]
    list3 = ["three", "four"]
    assert jaccard_similarity(list1, list2) == 1
    assert jaccard_similarity(list1, list3) == 0


def test_cosine_sim_vectors():
    vector1 = np.array([1, 1, 1])
    vector2 = np.array([1, 1, 1])
    assert cosine_sim_vectors(vector1, vector2) >= 1


def test_score_failures():
    failures = list(itertools.permutations(["i am failure 1", "i am failure 2", "i am failure 3", "i am failure 4"], 2))
    sm_ratios, coss, jaccards, jaros, levens = score_failures(failures)
    sum_sm = np.sum(sm_ratios)
    sum_coss = np.sum(coss)
    sum_jaccard = np.sum(jaccards)
    sum_jaros = np.sum(jaros)
    sum_levens = np.sum(levens)
    assert sum_sm > 0
    assert sum_coss > 0
    assert sum_jaccard > 0
    assert sum_jaros > 0
    assert sum_levens > 0


def test_invalid_path():
    with pytest.raises(IOError):
        run("not/here")


def test_console_output(capsys):
    run(str(XUNIT_FILES_DIR))
    captured = capsys.readouterr()
    assert EXPECTED_OUTPUT_START in captured.out
    assert EXPECTED_OUTPUT_END in captured.out
    assert "test_me.py:6: AssertionError\n" in captured.out


def test_no_failures(capsys):
    with pytest.raises(SystemExit):
        with tempfile.TemporaryDirectory() as temp_folder:
            shutil.copy(PASS_01_FILE_PATH, Path(temp_folder) / PASS_01_FILE_NAME)
            run(temp_folder)
            captured = capsys.readouterr()
            assert captured.out == "NO FAILURES FOUND"

    with pytest.raises(SystemExit):
        with tempfile.TemporaryDirectory() as temp_folder:
            run(temp_folder)
            captured = capsys.readouterr()
            assert captured.out == "NO FAILURES FOUND"


def test_finding_files():
    with tempfile.TemporaryDirectory() as temp_folder:
        folder_match_filter_patters = Path(temp_folder) / "xmlxml"
        folder_match_filter_patters.mkdir(parents=True)
        shutil.copy(PASS_01_FILE_PATH, folder_match_filter_patters / PASS_01_FILE_NAME)
        shutil.copy(FAIL_01_FILE_PATH, folder_match_filter_patters / FAIL_01_FILE_NAME)
        shutil.copy(FAIL_02_FILE_PATH, folder_match_filter_patters / FAIL_02_FILE_NAME)
        run(temp_folder)
