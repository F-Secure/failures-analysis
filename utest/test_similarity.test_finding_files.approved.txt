============== FAILURE START =================
def test_02():
>       assert False
E       assert False


tests\test_me.py:6: AssertionError
============== FAILURE END =================
                                         cos
suitename2    testname2 filename2           
tests.test_me test_02   failing_01_.xml  1.0
                        failing_02_.xml  1.0
